### ISSUE TITLE
#9587 exposes another error in the solver?

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/ghc#119]
I tried the code example in [#9587](https://ghc.haskell.org/trac/ghc/ticket/9587) in hope it will work thanks to injectivity:

``` haskell
{-# LANGUAGE NoMonomorphismRestriction #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE ScopedTypeVariables #-}

module T9587 where

-- Injectivity annotation added to hopefully solve the reported problem
type family Arr (repr :: * -> *) (a :: *) (b :: *) = (r :: *) | r -> repr a b

class ESymantics repr where
    int :: Int  -> repr Int
    add :: repr Int  -> repr Int -> repr Int

    lam :: (repr a -> repr b) -> repr (Arr repr a b)
    app :: repr (Arr repr a b) -> repr a -> repr b

te4 = let c3 = lam (\f -> lam (\x -> f `app` (f `app` (f `app` x))))
      in (c3 `app` (lam (\x -> x `add` int 14))) `app` (int 0)

```

But alas compilation ends with an error:

```
Couldn't match type ‘Arr repr (Arr repr a a) (Arr repr a a)’
               with ‘Arr repr (Arr repr a0 a0) (Arr repr a0 a0)’
NB: ‘Arr’ is a type function, and may not be injective
The type variable ‘a0’ is ambiguous
Expected type: repr (Arr repr (Arr repr a a) (Arr repr a a))
  Actual type: repr (Arr repr (Arr repr a0 a0) (Arr repr a0 a0))
When checking that ‘c3’ has the inferred type
  c3 :: forall (repr :: * -> *) a.
        ESymantics repr =>
        repr (Arr repr (Arr repr a a) (Arr repr a a))
Probable cause: the inferred type is ambiguous
In the expression:
  let c3 = lam (\ f -> lam (\ x -> ...))
  in (c3 `app` (lam (\ x -> x `add` int 14))) `app` (int 0)
In an equation for ‘te4’:
    te4
      = let c3 = lam (\ f -> ...)
        in (c3 `app` (lam (\ x -> x `add` int 14))) `app` (int 0)
```

This error message is better than without injectivity declaration (better = some type ambiguities seem to have been resolved), but we should do better.

**Comment by jstolarek** [comment]
This ticket also is another example of #107

**Comment by goldfirere** [comment]
I would want input from @simonpj before continuing here. Here's a summary about what is going on:

Here's where we start:

```
        [W] cobox_aoQ :: Arr
                           repr_aoN[tau:3]
                           (Arr repr_aoN[tau:3] a_aoO[tau:3] a_aoO[tau:3])
                           (Arr repr_aoN[tau:3] a_aoO[tau:3] a_aoO[tau:3])
                         ~ Arr
                             repr_aoK[sk]
                             (Arr repr_aoK[sk] a_aoL[sk] a_aoL[sk])
                             (Arr repr_aoK[sk] a_aoL[sk] a_aoL[sk]) (CNonCanonical)}
```

Flattening soon brings us to

```
  inerts = Equalities: [W] cobox_ap1 :: s_aoV[fuv:0]
                                        ~ s_aoZ[fuv:0] (CTyEqCan)
updWorkList
  WL {Funeqs = [W] cobox_aoY :: Arr repr_aoK[sk] a_aoL[sk] a_aoL[sk]
                                ~ s_aoX[fuv:0] (CFunEqCan)
               [W] cobox_ap0 :: Arr repr_aoK[sk] s_aoX[fuv:0] s_aoX[fuv:0]
                                ~ s_aoZ[fuv:0] (CFunEqCan)
               [W] cobox_aoU :: Arr repr_aoK[sk] a_aoO[tau:3] a_aoO[tau:3]
                                ~ s_aoT[fuv:0] (CFunEqCan)
               [W] cobox_aoW :: Arr repr_aoK[sk] s_aoT[fuv:0] s_aoT[fuv:0]
                                ~ s_aoV[fuv:0] (CFunEqCan)
```

`aoY`, `ap0`, and `aoU` quickly get worked on, nothing happens, and then these, too, become inert:

```
Start solver pipeline {
  work item =  [W] cobox_aoW :: Arr
                                  repr_aoK[sk] s_aoT[fuv:0] s_aoT[fuv:0]
                                ~ s_aoV[fuv:0] (CFunEqCan)
  inerts    =  Equalities: [W] cobox_ap1 :: s_aoV[fuv:0]
                                            ~ s_aoZ[fuv:0] (CTyEqCan)
               Type-function equalities: [W] cobox_aoY :: Arr
                                                            repr_aoK[sk] a_aoL[sk] a_aoL[sk]
                                                          ~ s_aoX[fuv:0] (CFunEqCan)
                                         [W] cobox_aoU :: Arr repr_aoK[sk] a_aoO[tau:3] a_aoO[tau:3]
                                                          ~ s_aoT[fuv:0] (CFunEqCan)
                                         [W] cobox_ap0 :: Arr repr_aoK[sk] s_aoX[fuv:0] s_aoX[fuv:0]
                                                          ~ s_aoZ[fuv:0] (CFunEqCan)
```

When `aoW` goes through, we emit new deriveds, according to the injectivity annotation on `Arr`:

```
runStage interact with inerts {
  workitem   =  [W] cobox_aoW :: Arr
                                   repr_aoK[sk] s_aoT[fuv:0] s_aoT[fuv:0]
                                 ~ s_aoV[fuv:0] (CFunEqCan)
interactFunEq improvements: 
  Eqns: [repr_aoK[sk] ~ repr_aoK[sk], s_aoT[fuv:0] ~ s_aoX[fuv:0],
         s_aoT[fuv:0] ~ s_aoX[fuv:0]]
  Candidates: [[W] cobox_aoY :: Arr repr_aoK[sk] a_aoL[sk] a_aoL[sk]
                                ~ s_aoX[fuv:0] (CFunEqCan),
               [W] cobox_aoU :: Arr repr_aoK[sk] a_aoO[tau:3] a_aoO[tau:3]
                                ~ s_aoT[fuv:0] (CFunEqCan),
               [W] cobox_ap0 :: Arr repr_aoK[sk] s_aoX[fuv:0] s_aoX[fuv:0]
                                ~ s_aoZ[fuv:0] (CFunEqCan)]
  TvEqs: [aoV :-> [[W] cobox_ap1 :: s_aoV[fuv:0]
                                    ~ s_aoZ[fuv:0] (CTyEqCan)]]
Emitting [D] [D] _ :: repr_aoK[sk] ~ repr_aoK[sk]
Emitting [D] [D] _ :: s_aoT[fuv:0] ~ s_aoX[fuv:0]
Emitting [D] [D] _ :: s_aoT[fuv:0] ~ s_aoX[fuv:0]
end stage interact with inerts }
```

So far, so good. The problem is that GHC doesn't recognize the `s_aoT ~ s_aoX` equation can actually lead to further improvement via injectivity annotations. This is seen as a straightforward `CTyVarEq`, and no type function magic happens. Note that `s_aoT` is the fuv for `Arr aoK aoO aoO`
and `s_aoX` is the fuv for `Arr aoK aoL aoL`. It's not hard to see that GHC should generated `aoO ~ aoL`, but this never happens. I wonder if a `CTyVarEq` between two fuvs is better represented as two `CFunEq`s. But I would imagine this would have already cropped up elsewhere -- hence my desire for Simon's input.

**Comment by jstolarek** [comment]
Aside from the original problem there might be one more issue in here:

```
Emitting [D] [D] _ :: s_aoT[fuv:0] ~ s_aoX[fuv:0]
Emitting [D] [D] _ :: s_aoT[fuv:0] ~ s_aoX[fuv:0]
```

Why is that emitted twice? I've already had a bug like this in my implementation of injectivity. I fixed it but there might be another one.

**Comment by goldfirere** [comment]
That's not problematic. It comes from the fact that we're dealing with `Arr xxx s_aoT s_aoT ~ Arr xxx s_aoX s_aoX`. So a naive algorithm would emit two Deriveds. A more advanced algorithm is possible, but I don't think it's worth the trouble, frankly. There is some caching that helps to reduce these, but only inerts are looked up. Right after a Derived is emitted, it's not yet inert and so the cache lookup fails. Fixing this is something to worry about later.

**Comment by goldfirere** [comment]
See a comment from Simon on #142, which addresses this bug.

**Comment by jstolarek** [comment]
(Replying to Simon's comment on #142)

> In short, in TcSMonad.kickOutRewritable (assuming you are up to date with HEAD), in the line
> 
> ```
> (feqs_out, feqs_in) = partitionFunEqs kick_out_ct funeqmap
> ```
> 
> you want to have a more specialised kick_out function, one that kicks it out if the RHS of the CFunEqCan is the same as the new_tv.

Sounds like a rebase would be needed. I tend to do upstream rebases. Richard, would this interfere with anything you are currently doing on my branch? Or perhaps we don't need a rebase? I suspect that might take a day or so and we don't have too much time before the deadline.

**Comment by simonpj** [comment]
Sounds like a rebase would be needed

I’m not sure.  But yes, the more up to date you are with HEAD the easier it is.

I’d like to be able to build this etc.  How do I do that? Is it in the main GHC repo, or just on Github?  If the latter, how do I pull from it? (I’m sure it’s easy, just have not done it recently.)  Which branch?  Etc

Simon

**Comment by jstolarek** [comment]
My branch is only on github: https://github.com/jstolarek/ghc/tree/T6018-injective-type-families

You should be able to get it like this:

```
git remote add jstolarek https://github.com/jstolarek/ghc
git fetch jstolarek
cd utils/haddock
git remote add jstolarek https://github.com/jstolarek/haddock
git fetch jstolarek
cd ../..
git checkout -b T6018-injective-type-families jstolarek/T6018-injective-type-families
git submodule update
```

My patch changes format of interface files, so you will need a full rebuild.

**Comment by goldfirere** [comment]
Rebase all you like. I do the same and know how to deal with the sequelae.

**Comment by jstolarek** [comment]
In that case I'll do a rebase this evening or tomorrow.

**Comment by jstolarek** [comment]
I applied the patch Simon mailed earlier:

```
--- a/compiler/typecheck/TcInteract.hs
+++ b/compiler/typecheck/TcInteract.hs
@@ -1115,8 +1115,8 @@ kickOutRewritable :: CtFlavour    -- Flavour of the equality that is
                   -> TcTyVar      -- The new equality is tv ~ ty
                   -> TcS Int
kickOutRewritable new_flavour new_eq_rel new_tv
-  | not ((new_flavour, new_eq_rel) `eqCanRewriteFR` (new_flavour, new_eq_rel))
-  = return 0  -- If new_flavour can't rewrite itself, it can't rewrite
+--  | not ((new_flavour, new_eq_rel) `eqCanRewriteFR` (new_flavour, new_eq_rel))
+--  = return 0  -- If new_flavour can't rewrite itself, it can't rewrite
               -- anything else, so no need to kick out anything
               -- This is a common case: wanteds can't rewrite wanteds
@@ -1159,7 +1159,7 @@ kick_out new_flavour new_eq_rel new_tv (IC { inert_eqs      = tv_eqs
                     , wl_implics = emptyBag }
     (tv_eqs_out, tv_eqs_in) = foldVarEnv kick_out_eqs ([], emptyVarEnv) tv_eqs
-    (feqs_out,   feqs_in)   = partitionFunEqs  kick_out_ct funeqmap
+    (feqs_out,   feqs_in)   = partitionFunEqs  kick_out_fe funeqmap
     (dicts_out,  dicts_in)  = partitionDicts   kick_out_ct dictmap
     (irs_out,    irs_in)    = partitionBag     kick_out_irred irreds
     (insols_out, insols_in) = partitionBag     kick_out_ct    insols
@@ -1168,6 +1168,11 @@ kick_out new_flavour new_eq_rel new_tv (IC { inert_eqs      = tv_eqs
     can_rewrite :: CtEvidence -> Bool
     can_rewrite = ((new_flavour, new_eq_rel) `eqCanRewriteFR`) . ctEvFlavourRole
+    kick_out_fe :: Ct -> Bool
+    kick_out_fe (CFunEqCan { cc_ev = ev, cc_fsk = fsk })
+      =  kick_out_ctev ev || fsk == new_tv
+    kick_out_fe _ = False  -- Can't happen
+
     kick_out_ct :: Ct -> Bool
     kick_out_ct ct = kick_out_ctev (ctEvidence ct)
```

The only difference is that now this is in `TcSMonad` and not `TcInteract`. To my surprise it does not fix this ticket -- ie. 9587 still does not work -- but it fixes #112. I'm puzzled.

**Comment by goldfirere** [comment]
Why comment out the call to `eqCanRewriteFR`? I don't think that's changed here.

I don't have much insight as to why this doesn't solve this bug, but I haven't yet rebuilt and such.

**Comment by jstolarek** [comment]
> Why comment out the call to `eqCanRewriteFR`? I don't think that's changed here.

I have no idea what it does, but from experimenting I can tell that restoring call to `eqCanRewriteFR`:
1. still does not fix this ticket
2. restores issues with numeric literals (#112)

**Comment by jstolarek** [comment]
Fixed by Simon in 5d8a71bc514f3128d18a2d6e4ec5cf29acb68e6c
